### PR TITLE
scripts: version-literal checker — forbid hardcoded pfSense/FreeBSD version tokens outside the matrix

### LINKED ISSUE #922: scripts: version-literal checker — forbid hardcoded pfSense/FreeBSD version tokens outside the matrix files
Follow-up from the 2026-07-07 workflow audit. The "cover ALL supported versions/editions" rule (CLAUDE.md upstream-version ladder; ADR-20's varver-not-ABI rule) has no mechanical backstop: hardcoded version/edition literals ship in code, comments, and workflows, and each one silently narrows scope when the matrix moves (observed in the #900–#909 wave and in commits 25d29e5d/94b499fc's sweep gaps).

**Proposal.** Add `scripts/check_version_literals.py` in the mold of `check_appliance_python.py` (pre-commit + CI, unit-tested):

1. Scan tracked `src/**`, `scripts/**`, `.github/workflows/**` for pfSense/FreeBSD version-shaped literals: `2.8`/`2.9` CE versions, `25.x`/`26.x` Plus versions, `FreeBSD:1[56]` ABI strings, `php8x`/`py31x` flavor tokens, `ce-X.Y`/`plus-XX.YY` varvers.
2. Allowlist the sanctioned sources of truth (the ci-metadata version matrix, `docs/misc/pfSense_versions.md`, `scripts/misc/install_deps_*` filenames, and an explicit inline-comment escape (`# version-literal-ok: <reason>`).
3. Violation → CI-blocking error naming the file:line and pointing at the matrix file to read the value from at runtime/generation time instead.

**Constraints.**

- Pure detection function + unit tests (`tests/test_version_literal_check.py`), wired into `.githooks/pre-commit` (path-scoped) and `test.yml`, same as the existing policy scripts.
- Expect a noisy first run: the initial PR should either seed the allowlist with the current legitimate sites (each with the escape comment and a reason) or land the checker in warn-only mode with a follow-up to flip it blocking — implementer's judgment, but the end state is blocking.
- Branch coverage: literal in src (fail), literal in allowlisted file (pass), escape comment (pass), matrix file itself (pass), workflow file (fail).

**Why.** Scope-subset via hardcoding was the single most common defect class in the audited window; the version matrix already exists as machine-readable truth, so restating it as literals is never necessary.

### PR DESCRIPTION
**Pull request by andrebrait**
## What

Adds `scripts/check_version_literals.py`, a preventative pre-commit + CI policy checker (in the mold of `check_appliance_python.py`) that forbids restating supported pfSense/FreeBSD version tokens as **hardcoded literals** — they must be read from the ci-metadata matrix (`supported-versions.json` via `read-version-matrix.sh`) at runtime, so a script/workflow can't silently drift when a version is added or dropped.

Fixes #922.

## Scope — values only, deliberately low false-positive

The user chose **values-only** enforcement. The checker flags a version token (CE `2.8`/`2.9`, Plus `25.NN`/`26.NN`, `FreeBSD:15`/`16` ABI, `php8x`/`py31x` flavors, `ce-X.Y`/`plus-NN.NN` varvers) **only when it stands alone as a value** — the entire content of a quoted literal (`"py311"`, `'FreeBSD:15:amd64'`) or the whole RHS of a `key=value`/`key: value` assignment. A token embedded in prose, `--help` text, a `#` comment, or a docstring stays clean, as does a flavor inside a package name (`py311-sqlite3`).

Excluded paths: `*.md`, `scripts/misc/install_deps_*` (real FreeBSD package names), and the checker + its own test (self-reference). Inline escape: `# version-literal-ok: <reason>`.

## Wiring + seed (lands blocking)

- **`.githooks/pre-commit`**: a new `staged_yaml` class + a version-literals gate that runs when any scan-root language (py/php/sh/yaml) is staged.
- **`.github/workflows/test.yml`**: a blocking policy-check step beside the appliance-python check.
- **Seeded the 17 pre-existing value-position hits** so the tree passes clean: 15 dev-tooling / `workflow_dispatch` fallback defaults get an inline escape (the real value is passed from the matrix at runtime — verified each default is overridable); the `test-version-matrix` sanity `jq` pin is escaped (jq accepts the `#` comment — verified); the `version-tracker` help-text example is de-quoted so it reads as prose.

## Tests

`tests/test_version_literal_check.py` — 22 discriminating tests: every token shape flagged as a value; every embedded/prose/comment/docstring/package-name form kept clean; path exclusions; escape suppression; a vacuity guard (`"3.11"`, `"8.3"`, `FreeBSD:14`/`17` stay clean). Non-vacuity proven by weakening the value-position rule → 7 discriminators go red → restore → green.

## Notes

- Documented ceiling (`ponytail:` comment): a version token inside a multi-line YAML folded/literal block scalar isn't recognized as prose — only `#` comments and Python docstrings are tracked; `version-literal-ok` covers the single such site today.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added automated checks to catch hardcoded version values in tracked files and CI workflows.
  * Expanded workflow and script guidance for intentionally allowed version values.

* **Bug Fixes**
  * Reduced the risk of version drift by encouraging version values to be sourced dynamically.
  * Improved handling of version defaults and examples across build and smoke workflows.

* **Tests**
  * Added coverage for the new version-literal validation behavior and related edge cases.

### CODE REVIEW

File: scripts/check_version_literals.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+SCOPE (deliberately low false-positive: VALUES only, never prose)
+-------------------------------------------------------------------
+* Scans tracked files under ``src/``, ``scripts/``, and ``.github/workflows/``
+  (production code, dev/CI tooling, and workflow YAML — everywhere a version
+  literal could plausibly be pasted).
+* Excludes: any ``*.md`` path (docs describe value *formats*, not enforce
+  them — the user chose values-only enforcement); ``scripts/misc/install_deps_*``
+  (real FreeBSD package names such as ``py311-sqlite3`` legitimately hardcode
+  a flavor there — an intentional allowlist); this file and its own test
+  (they define/contain the patterns being matched, so scanning them is
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply scripts/check_version_literals.py:33]: Applied in e486f172 — reworded the docstring (and the inline comment on `_is_excluded`) to describe the exclusion as an `install_deps_*` **filename** match rather than implying a `scripts/misc/` path constraint. The filename match is the intended allowlist (issue #922 lists "`scripts/misc/install_deps_*` filenames"), so the behavior is correct and the docs now agree with it.

File: tests/test_version_literal_check.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+# --- Row 8: unquoted standalone RHS -> FLAGGED (shell and YAML) -----------
+
+
+def test_flags_unquoted_standalone_rhs(tmp_path: Path) -> None:
+    shell_line = f"ABI={_FREEBSD15}\n"
+    yaml_line = f"default: {_PY311}\n"
+    assert len(_find(tmp_path, shell_line)) == 1, "unquoted shell assignment must be flagged"
+    assert len(_find(tmp_path, yaml_line)) == 1, "unquoted YAML assignment must be flagged"
+
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply tests/test_version_literal_check.py:118]: Applied in e486f172 — added `test_export_readonly_prefixed_unquoted_assignment_flagged` covering `export ABI=…` and `readonly PYF=…`, alongside the regex fix below.

File: scripts/check_version_literals.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+# matching if hardcoded dependency names spread beyond the allowlisted file.
+#
+# ponytail: a quoted illustrative example inside a multi-line YAML folded/
+# literal scalar (`description: >` ... "e.g. \"2.8\"" on a continuation line)
+# is not recognised as prose -- only `#` comments and Python triple-quoted
+# docstrings are tracked. Upgrade to a YAML block-scalar tracker (indentation-
+# based, mirroring _prose_line_flags's docstring state machine) if more than
+# this one line ever needs it; today `version-literal-ok` covers it.
+_FULL_VALUE_RE = re.compile("^(?:" + "|".join(_TOKEN_ALTERNATIVES) + ")$")
+_ASSIGNMENT_RE = re.compile(r"^\s*[\w.-]+\s*[:=]\s*(?:" + "|".join(_TOKEN_ALTERNATIVES) + r")\s*$")
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply scripts/check_version_literals.py]: Confirmed and fixed in e486f172. Empirically reproduced first: `export ABI=FreeBSD:15:amd64` (unquoted) was missed while the quoted form was caught by the quoted-literal path. `_ASSIGNMENT_RE` now allows an optional `export `/`readonly ` prefix, with a red→green test.

**Comment by andrebrait** [comment]
## Review summary — PR #937

Three review sources ran on this PR:

- **Claude Sonnet 5 adversarial review** — single agent at reasoning effort `xhigh` (the `review-fanout` multi-agent shape was attempted first but hit an args-plumbing error in this environment; the single `xhigh` adversarial pass was used instead). It ran the checker, the full suite, `ruff`/`mypy`/`shellcheck`, YAML-parsed every touched workflow, verified every escape-comment claim against the actual calling code, and hand-built ~20 adversarial inputs against the parser.
- **GitHub Copilot** — 3 inline findings.
- **CodeRabbit** — acknowledged, then rate-limited before it could review; the push of the review fixes re-triggers it.

Every finding was **applied** (`e486f172`); none skipped or deferred.

### Findings ledger

| # | Source | Severity | Finding | Resolution |
|---|--------|----------|---------|------------|
| F1 | Sonnet | **blocking** | `_prose_line_flags` treated any `"""`/`'''` line as prose in **all** file types, so a shell `VAR="""FreeBSD:15:amd64"""` (POSIX adjacent-quote concatenation → the literal) bypassed the gate. Latent (no such line in the tree) but a trivial bypass of a blocking gate. | **fixed@e486f172** — triple-quote docstring tracking scoped to `.py` only (`#` comments stay prose in every language). Red→green: `test_shell_triple_quoted_value_flagged` fails on the pre-fix checker, passes after; `.py` docstrings still exempt (`test_python_docstring_body_still_exempt`). |
| Copilot-3 | Copilot | valid | `_ASSIGNMENT_RE` missed `export`/`readonly`-prefixed **unquoted** assignments (`export ABI=FreeBSD:15:amd64`). | **fixed@e486f172** — optional `export`/`readonly` prefix added; reproduced first, then red→green test. |
| Copilot-1 | Copilot | nitpick | Docstring implied a `scripts/misc/` path constraint on the `install_deps_*` exclusion, but the code matches on filename. | **fixed@e486f172** — docs reworded to a filename match (matches the issue's `install_deps_*` intent). |
| Copilot-2 | Copilot | nitpick | No test for the `export ABI=…` shell form. | **fixed@e486f172** — `test_export_readonly_prefixed_unquoted_assignment_flagged` added. |
| F2 | Sonnet | nitpick | Unquoted-RHS detection misses compound statements (`A=x; B=y`) and unquoted YAML sequence items. None in the tree; spec scopes to `key: value`. | **fixed@e486f172** — documented as a `ponytail:` ceiling comment (a quoted token in any of these is still caught). |
| F3 | Sonnet | nitpick | The YAML-folded-scalar ceiling comment claimed `version-literal-ok` covered `version-tracker.yml`, but that line was actually fixed by **de-quoting**. | **fixed@e486f172** — comment corrected. |
| F4 | Sonnet | nitpick | The `php8[0-9]` token shape had no explicit test row. | **fixed@e486f172** — `test_flags_php_flavor_quoted_value` added. |
| F5 | Sonnet | nitpick | The failure message pointed at CLAUDE.md, which didn't document the policy. | **fixed@e486f172** — checker documented in CLAUDE.md's Linting section, like its siblings. |

Gates after the fixes: `check_version_literals.py` exits 0 over the tree, `ruff`/`ruff format`/`mypy tests/` clean, full pytest **2328 passed, 5 skipped**, `markdownlint` clean.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.
